### PR TITLE
design-docs: Add a story about specifying the distribution type

### PR DESCRIPTION
**Pull request by sschuberth**
Any of the checked boxes below indicate that I took action:

- [x] Reviewed the [Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md#contribution-workflow).
- [x] Signed the [Gradle CLA](http://gradle.org/contributor-license-agreement/).
- [x] Ensured that basic checks pass: `./gradlew quickCheck`

This PR contains the design docs proposal only. The PR with the non-trivial changes will follow once this PR is merged.

### CODE REVIEW

File: design-docs/done/build-initialisation.md [hunk 1]
```diff
@@ -166,3 +166,14 @@ This story adds a commandline property `--gradle-version` to the `wrapper` task
 
 * Running `gradle wrapper --gradle-version 1.6` generates valid `wrapper.properties` with correct URL.
 
+# Story: User specifies target gradle distribution type when generating the wrapper
+
+This story adds a command line option called `--gradle-type` to the `wrapper` task to specify the desired Gradle distribution type:
```
Comment (wolfs) [inline design-docs/done/build-initialisation.md]: How about calling the option `dist-type`?

Comment (sschuberth) [inline design-docs/done/build-initialisation.md]: I'd be OK with that. I just thought it should start with `--gradle` to be consistent with `--gradle-version`. So, do you want me to use `--dist-type` instead`?

File: design-docs/done/build-initialisation.md [hunk 2]
```diff
@@ -166,3 +166,14 @@ This story adds a commandline property `--gradle-version` to the `wrapper` task
...
+
+* Add `--gradle-type` command-line option to `wrapper`.
+
+Being able to set the distibution type to `all` instead of the default of `bin` is useful if the source code is required for IDEs to offer proper auto-completion for the DSL.
+
+## Test coverage
+
+* Running `gradle wrapper --gradle-version 2.13 --gradle-type all` generates valid `wrapper.properties` with the expected URL.
```
Comment (wolfs) [inline design-docs/done/build-initialisation.md]: There should be more test coverage:

 - What happens if there is no `--dist-type` command line option
 - What happens if the `--dist-type` in the command line does not exist

**Comment by wolfs** [comment]
Thank you very much for your pull request!

All in all the design doc looks good. I added some inline comments. Please address those in order to get this merged.

File: design-docs/done/build-initialisation.md (hunk 1 above)
Comment (wolfs) [inline design-docs/done/build-initialisation.md]: Yes, please use `dist-type`. Thank you!

File: design-docs/done/build-initialisation.md [hunk 3]
```diff
@@ -166,3 +166,14 @@ This story adds a commandline property `--gradle-version` to the `wrapper` task
 
 * Running `gradle wrapper --gradle-version 1.6` generates valid `wrapper.properties` with correct URL.
 
+# Story: User specifies target gradle distribution type when generating the wrapper
+
+This story adds a command line option called `--gradle-type` to the `wrapper` task to specify the desired Gradle distribution type:
+
+* Add `--gradle-type` command-line option to `wrapper`.
```
Comment (wolfs) [inline design-docs/done/build-initialisation.md]: Also add "update documentation" to the implementation plan.

**Comment by sschuberth** [comment]
@wolfs I've updated the PR to address your comments.

**Comment by wolfs** [comment]
Thank you for your work! I merged the pull request.

Are you planning on implementing this, soon?

**Comment by sschuberth** [comment]
It's already implemented :-) I just wanted to get the design docs accepted first. Will file a PR with the implementation soon.

**Comment by bmuschko** [comment]
@wolfs Can you add the test case "Running `gradle wrapper --gradle-version 2.13 --dist-type bin` generates valid `wrapper.properties` with `bin` distribution URL."? Otherwise LGTM.

**Comment by sschuberth** [comment]
PS, the

    * Running `gradle wrapper --gradle-version 2.13` defaults to `bin` as the distribution type.

test case is already there, in the form of "generated wrapper scripts for given version from command-line". So I guess I do not need to add that again?

**Comment by sschuberth** [comment]
Also, I just check how the `wrapper` task behaves if an invalid version (like `9.87`) is specified. It just literally copies the version number to `gradle-wrapper.properties` without throwing an error. Which makes sense as checking for a valid version would require an Internet connection. So am I correct to assume that for an invalid distribution type it should behave the same, and that the

    * Running `gradle wrapper --gradle-version 2.13 --dist-type invalid_distribution_name` generates an error.

test from the design docs should be omitted?

**Comment by bmuschko** [comment]
@sschuberth We'll add that test case in the spec. Don't worry about it. I'd like to spell it out directly as we need to make sure that `--dist-type bin` produces the same result as the default behavior. Regarding the version number. We don't check the version as you can create your own custom Gradle distribution with a custom version number. For `--dist-type` it's slightly different. We only want to accept `bin` or `all` distributions at the moment.

**Comment by sschuberth** [comment]
> For `--dist-type` it's slightly different. We only want to accept `bin` or `all` distributions at the moment.

Alright. I'm now throwing an `CommandLineArgumentException` if `dist-type` is neither `bin` nor `all`. Is that the correct exception to throw?

Also, I've added a test that check for that exception to be thrown, but running it I get an `WrongExceptionThrownError`:

    org.gradle.integtests.WrapperGenerationIntegrationTest > generated wrapper scripts for given distribution type from command-line FAILED
        org.spockframework.runtime.WrongExceptionThrownError at WrapperGenerationIntegrationTest.groovy:76
            Caused by: org.gradle.integtests.fixtures.executer.UnexpectedBuildFailure at WrapperGenerationIntegrationTest.groovy:73
                Caused by: org.gradle.internal.exceptions.LocationAwareException at WrapperGenerationIntegrationTest.groovy:73
                    Caused by: org.gradle.cli.CommandLineArgumentException at WrapperGenerationIntegrationTest.groovy:73

Howq can I make Spock look at the original exception being thrown? Or is there another way to fix this?

**Comment by wolfs** [comment]
@sschuberth: Feel free to open a PR then we can easier have a discussion about the problems you are encountering.

Regarding your problem: I suppose you are writing an integration test? Then you should check that the build fails and check the output.

**Pull request merged**